File: solutions/188.best-time-to-buy-and-sell-stock-iv.js
/*
 * @lc app=leetcode id=188 lang=javascript
 *
 * [188] Best Time to Buy and Sell Stock IV
 */

// @lc code=start
/**
 * @param {number} k
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (k, prices) {
  const len = prices.length;
  if (len == 0) return 0;

  const dp = Array.from({ length: k + 1 }, () =>
    Array.from({ length: len }, () => ({}))
  );

  // 初始化第零天状态
  for (let act = 0; act <= k; ++act) {
    dp[act][0].no_hold = 0;
    dp[act][0].hold = -prices[0];
  }

  // 初始化0次交易。 0次交易是不合理的存在，所以dp长度是 k+1
  for (let day = 1; day < len; day++) {
    dp[0][day].no_hold = 0;
    dp[0][day].hold = Number.MIN_VALUE;
  }

  for (let act = 1; act <= k; ++act) {
    for (let day = 1; day < len; ++day) {
      /**
       * 不持有股票分两种情况
       * 1：前一天不持有，今天无操作
       * 2: 前一天持有，今天卖了
       * 因为无买入操作，所以是在同一个act维度
       */
      dp[act][day].no_hold = Math.max(
        dp[act][day - 1].no_hold,
        dp[act][day - 1].hold + prices[day]
      );
      /**
       * 持有股票分两种情况
       * 1：前一天不持有，今天买入， 买入是一个新的act，所以前一天的交易是act-1维度
       * 2: 前一天持有，今天无操作， 不耗费act，所以是same act 维度
       */
      dp[act][day].hold = Math.max(
        dp[act - 1][day - 1].no_hold - prices[day],
        dp[act][day - 1].hold
      );
    }
  }

  return dp[k][len - 1].no_hold;
};
// @lc code=end
maxProfit(1, [1, 2]);

File: solutions/188.best-time-to-buy-and-sell-stock-iv.rb
# @param {Integer} k
# @param {Integer[]} prices
# @return {Integer}
def max_profit(k, prices)
  len = prices.length
  dp = Array.new(len) do
    Array.new(k + 1) { State.new }
  end

  (0...k + 1).each do |act| # 0次交易只是概念上的存在，是初始化值，非合理存在，几交易次数应该是: 1次，2次，... k次, 所以 0...k+1
    dp[0][act].no_hold = 0
    dp[0][act].hold = -prices[0] # 第‘0’天 无论执行多少次交易(0...k+1),都是花掉这天买股票的钱。
  end

  (0...len).each do |day|
    dp[day][0].no_hold = 0
    dp[day][0].hold = -1e10
  end

  (1...len).each do |day|
    (1...k + 1).each do |act|
      today_price = prices[day]
      dp[day][act].no_hold = max(dp[day - 1][act].no_hold, dp[day - 1][act].hold + today_price)
      dp[day][act].hold = max(dp[day-1][act - 1].no_hold - today_price, dp[day - 1][act].hold)  # 根据👆定义，只有[买入]才消耗act
    end
  end

  dp[len - 1][k].no_hold
end

class State
  attr_accessor :hold, :no_hold
end

def max(*vals) = vals.max
